LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/WebSocketWriterRegistry.cs

```diff
@@ -0,0 +1,182 @@
+        /// </summary>
+        /// <remarks>
+        /// This method is idempotent.
+        /// </remarks>
+        public IWebSocketWriterHandle<TMessage> Register(TKey key, IWebSocketWriter<TMessage> socketWriter)
+        {
+            lock (this.sockets)
+            {
+                Handle handle;
+
+                if (this.sockets.TryGetValue(key, out var current))
+                {
+                    handle = current.Handle;
+                    if (socketWriter == current.Object)
+                        return handle;
+                }
+                else
+                {
+                    handle = new Handle(this, key);
+                }
+
+                this.sockets[key] = (socketWriter, handle);
+                return handle;
+            }
+        }
+
+        /// <summary>
+        /// Deregisters any socket writer associated with a key, return the associated writer if one
+        /// was previously registered.
+        /// </summary>
+        public IWebSocketWriter<TMessage>? Deregister(TKey key)
+        {
+            lock (this.sockets)
+            {
+                if (this.sockets.TryGetValue(key, out var current))
+                {
+                    _ = this.sockets.Remove(key);
+                    return current.Object;
+                }
+                else
+                {
+                    return null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Sends a message to the socket writer associated with the given key.
+        /// </summary>
+        /// <remarks>
+        /// If the send fails because the socket has been closed prematurely then the socket writer
+        /// is unregistered before this method returns.
+        /// </remarks>
+        private async ValueTask SendAsync(TKey key, TMessage message, CancellationToken cancellationToken)
+        {
+            IWebSocketWriter<TMessage> socketWriter;
+
+            lock (this.sockets)
+                socketWriter = this.sockets[key].Object;
+
+            try
+            {
+                await socketWriter.SendAsync(message, cancellationToken).ConfigureAwait(false);
+            }
+            catch (WebSocketException ex) when (ex.WebSocketErrorCode == WebSocketError.ConnectionClosedPrematurely)
+            {
+                _ = Deregister(key);
+            }
+        }
+
+        /// <summary>
+        /// Removes closed sockets at default interval periods until cancellation is requested.
+        /// </summary>
+        public Task RunPrunerAsync(CancellationToken cancellationToken) =>
+            RunPrunerAsync(WebSocketWriterRegistry.DefaultPruningInterval, cancellationToken);
+
+        /// <summary>
+        /// Removes closed sockets at a given a interval period until cancellation is requested.
+        /// </summary>
+        public Task RunPrunerAsync(TimeSpan interval, CancellationToken cancellationToken)
```

Comment:
I don't see this is being used anywhere except for test?

Reply:
True. This was added in anticipation of being used, but then when I got to integrating the registry into the `LnsProtocolMessageProcessor`, it didn't seem absolutely necessary. If the socket gets closed then everything should clean up in an orderly fashion. If it's closed abnormally then an error will result during the next attempt to read or send a message and either will cause a deregistration. This could come under YAGNI so happy to remove for now and bring it back if and when we need it.